### DISCUSSION TITLE
Create variable stiffness beam

### DISCUSSION THREAD

**Discussion by Yangon0620** [sofa-framework/sofa#4952] [BeamAdapter] [answered]
I am creating a beam structure with different stiffness sections in Sofa 23.06, and I would like to use the `InterventionalRadiologyController ` component to control the Extension and Contraction of the beam. I tried using the
 `IndexValueMapper` component to modify the default YoungModulus and Poissonratio values in the `WireBeamInterpolation`  . Although the values were changed, there was no change in the stiffness of the beam in practice. I guess it's because the   `WireBeamInterpolation`  inherits the YoungModulus from the `WireRestShape`  component. I would like to ask, how can I create a beam segment with a different stiffness?

```xml
<?xml version="1.0"?>
<Node name="root" gravity="0 -9.81 0" dt="0.01" bbox="0 0 0 10 10 10">
    <RequiredPlugin name="BeamAdapter"/> <!-- Needed to use components [WireBeamInterpolation WireRestShape] -->
    <RequiredPlugin name="Sofa.Component.Constraint.Projective"/> <!-- Needed to use components [FixedConstraint] -->
    <RequiredPlugin name="Sofa.Component.LinearSolver.Direct"/> <!-- Needed to use components [BTDLinearSolver] -->
    <RequiredPlugin name="Sofa.Component.ODESolver.Backward"/> <!-- Needed to use components [EulerImplicitSolver] -->
    <RequiredPlugin name="Sofa.Component.SolidMechanics.Spring"/> <!-- Needed to use components [RestShapeSpringsForceField] -->
    <RequiredPlugin name="Sofa.Component.StateContainer"/> <!-- Needed to use components [MechanicalObject] -->
    <RequiredPlugin name="Sofa.Component.Topology.Container.Dynamic"/> <!-- Needed to use components [EdgeSetGeometryAlgorithms EdgeSetTopologyContainer EdgeSetTopologyModifier] -->
    <RequiredPlugin name="Sofa.Component.Topology.Container.Grid"/> <!-- Needed to use components [RegularGridTopology] -->
    <RequiredPlugin name="Sofa.Component.Visual"/> <!-- Needed to use components [VisualStyle] -->
    <VisualStyle displayFlags="showVisualModels showBehaviorModels showCollisionModels hideMappings showForceFields" />

    <DefaultAnimationLoop />
    <DefaultVisualManagerLoop />
    <Node name="EdgeTopology">
        <WireRestShape name="BeamRestShape" straightLength="60.0" length="40.0"
                                 numEdges="200" youngModulus="6000"
                                 spireDiameter="25" numEdgesCollis="50 10"
                                 printLog="1" template="Rigid3d" spireHeight="0.0"
                                 densityOfBeams="4 3" youngModulusExtremity="20000" />
        <EdgeSetTopologyContainer name="meshLinesBeam"/>
        <EdgeSetTopologyModifier name="Modifier"/> 
        <EdgeSetGeometryAlgorithms name="GeomAlgo" template="Rigid3d" /> 
        <MechanicalObject name="dofTopo2" template="Rigid3d" />
    </Node>  

    <Node name="BeamModel">
        <EulerImplicitSolver rayleighStiffness="0.2" rayleighMass="0.1" printLog="false" />
        <BTDLinearSolver verbose="0"/>
        <RegularGridTopology name="MeshLines" nx="8" ny="1" nz="1"   
                                    xmax="0.0" xmin="1.0" ymin="0" ymax="0" zmax="0" zmin="0"
                                    p0="0 0 0" drawEdges="1"/>
        <MechanicalObject template="Rigid3d" name="DOFs Container" ry="-90" />
        <IndexValueMapper name="tip_YoungModulus" indices = "4 5 6" value= "92000"/>
		<IndexValueMapper name="whole_beam_YoungModulus" indices = "0 1 2 3" value = "10000" inputValues = "@tip_YoungModulus.outputValues"/>

        <WireBeamInterpolation name="BeamInterpolation" WireRestShape="@../EdgeTopology/BeamRestShape" defaultYoungModulus = "@whole_beam_YoungModulus.outputValues" radius="0.9"/>

        <AdaptiveBeamForceFieldAndMass name="BeamForceField"  massDensity="0.00000155" interpolation="@BeamInterpolation"/>
        <InterventionalRadiologyController name="DeployController" template="Rigid3d" instruments="BeamInterpolation"
                                    startingPos="0 0 0 0 0 0 1" xtip="0 0 0" printLog="1" 
                                    rotationInstrument="0 0 0" step="0.5" speed="0.5" 
                                    listening="1" controlledInstrument="0"/>
        <FixedConstraint name="FixedConstraint" indices="0" />
        <RestShapeSpringsForceField name="RestSPForceField" points="@DeployController.indexFirstNode" angularStiffness="1e8" stiffness="1e8"/> 
        <ConstantForceField  indices = "7" forces = "0 0 0 1 50 50 " showArrowSize = "1"/>
    </Node>  
</Node>
```
![image](https://github.com/user-attachments/assets/8bc59938-42a8-44e5-823b-b38ab38e7565)

When I upgraded Sofa to version 24.06, if I used ` WireRestShape` to create at least 3 `RodxxxSections` , the beams created would exhibit very abnormal movement states.
And `WireRestShape`  can not create the Beam with two different `RodStraightSection`
![image](https://github.com/user-attachments/assets/623be7f6-852e-4c8e-99aa-b281088448f7)

So, I have no idea about how to create a variable stiffness beam in a proper way. Please help me with this, Thanks alot

**Comment by Yangon0620** [comment]
![image](https://github.com/user-attachments/assets/43a1a5f0-063c-4d52-ad3d-48d687d002f1)
Here is the first picture about modifying the defaultYoungModulus.

And here is the code about Rodxxxsection:
<?xml version="1.0"?>
<Node name="root" gravity="0 -9.81 0" dt="0.01" bbox="0 0 0 10 10 10">
<RequiredPlugin name="BeamAdapter"/> 
<RequiredPlugin name="Sofa.Component.Constraint.Projective"/> 
<RequiredPlugin name="Sofa.Component.LinearSolver.Direct"/> 
<RequiredPlugin name="Sofa.Component.ODESolver.Backward"/> 
<RequiredPlugin name="Sofa.Component.SolidMechanics.Spring"/> 
<RequiredPlugin name="Sofa.Component.StateContainer"/> 
<RequiredPlugin name="Sofa.Component.Topology.Container.Dynamic"/> 
<RequiredPlugin name="Sofa.Component.Topology.Container.Grid"/> 
<RequiredPlugin name="Sofa.Component.Visual"/> 
<VisualStyle displayFlags="showVisualModels showBehaviorModels showCollisionModels hideMappings showForceFields" />
<DefaultAnimationLoop />
<DefaultVisualManagerLoop />

<Node name="EdgeTopology">
    <RodStraightSection name="StraightSection" youngModulus="20000" radius="1" nbEdgesCollis="30" nbEdgesVisu="196" length="980.0"/>
    <RodSpireSection name="SpireSection" youngModulus="8000" radius="1" nbEdgesCollis="5" nbEdgesVisu="4" length="20.0" spireDiameter="25" spireHeight="0.0"/>
    <RodStraightSection name="StraightSection2" youngModulus="12000" radius="1" nbEdgesCollis="5" nbEdgesVisu="4" length="20.0" spireDiameter="10" spireHeight="10.0"/> 
    <WireRestShape template="Rigid3d" name="BeamRestShape" printLog="1" wireMaterials="@StraightSection @StraightSection2"/>    
    <EdgeSetTopologyContainer name="meshLinesBeam"/>
    <EdgeSetTopologyModifier name="Modifier"/> 
    <EdgeSetGeometryAlgorithms name="GeomAlgo" template="Rigid3d" /> 
    <MechanicalObject name="dofTopo2" template="Rigid3d" />
</Node>  

<Node name="BeamModel">
    <EulerImplicitSolver rayleighStiffness="0.2" rayleighMass="0.1" printLog="false" />
    <BTDLinearSolver verbose="0"/>
    <RegularGridTopology name="MeshLines" nx="60" ny="1" nz="1"
                                xmax="0.0" xmin="0.0" ymin="0" ymax="0" zmax="0" zmin="0"
                                p0="0 0 0" drawEdges="1"/>
    <MechanicalObject template="Rigid3d" name="DOFs Container" ry="-90" /> 
    <WireBeamInterpolation name="BeamInterpolation" WireRestShape="@../EdgeTopology/BeamRestShape" radius="0.9"/>
    <AdaptiveBeamForceFieldAndMass name="BeamForceField"  massDensity="0.00000155" interpolation="@BeamInterpolation"/>
    <InterventionalRadiologyController name="DeployController" template="Rigid3d" instruments="BeamInterpolation"
                                startingPos="0 0 0 0 0 0 1" xtip="0 0 0" printLog="1" 
                                rotationInstrument="0 0 0" step="0.5" speed="0.5" 
                                listening="1" controlledInstrument="0"/>
    <FixedConstraint name="FixedConstraint" indices="0" />
    <RestShapeSpringsForceField name="RestSPForceField" points="@DeployController.indexFirstNode" angularStiffness="1e8" stiffness="1e8"/> 
</Node>  
</Node>

**Reply by Yangon0620** [reply]
![image](https://github.com/user-attachments/assets/5fbdb1f5-ec6c-44e3-9a6b-fc101cc280ae)

**Comment by epernod** [comment, marked as answer]
Hi,

I suggest you update your SOFA and BeamAdapter version to at least v23.12 and even better v24.06.
Many changes have been done in BeamAdpater to use sections and define physical/geometrical parameters per section. Or you can use BeamInterpolation if in your case you do not need adaptive wire and in this component you can define the physical parameters per beam.

See for example this [PR](https://github.com/sofa-framework/BeamAdapter/pull/155)

Best regards
